In [1]:
import pandas as pd
import numpy as np
import polars as pl

from m4_feats_polars import *
from m5_sb_models import *
import warnings
warnings.filterwarnings("ignore")

In [2]:
# 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
data_path     = 'kaggle/input/linking-writing-processes-to-writing-quality/'
train_logs    = pl.scan_csv(f'{data_path}/train_logs.csv')
test_logs    = pl.scan_csv(f'{data_path}/test_logs.csv')
train_scores = pl.scan_csv(f'{data_path}/train_scores.csv')

# train_logs, test_logs = amend_event_id_order(train_logs, test_logs)     worsens

In [28]:
def essay_sent_words(df):
    AGGREGATIONS = ['count', 'mean', 'max', 'first', q1, 'median', q3, 'sum']
    df['sent'] = df['essay'].apply(lambda x: re.split('\\.|\\?|\\!',x))
    df = df.explode('sent')
    df['sent'] = df['sent'].apply(lambda x: x.replace('\n','').strip())
    df['sent_word_count'] = df['sent'].apply(lambda x: len(x.split(' ')))

    sent_agg_df = df[['id','sent_word_count']].groupby(['id']).agg(AGGREGATIONS)
    sent_agg_df.columns = ['_'.join(x) for x in sent_agg_df.columns]
    sent_agg_df['id'] = sent_agg_df.index
    sent_agg_df = sent_agg_df.reset_index(drop=True)
    sent_agg_df.drop(columns=["sent_word_count_count"], inplace=True)
    return sent_agg_df

def essay_sent_length(df):
    AGGREGATIONS = ['count', 'mean', 'min', 'max', 'first', 'last', q1, 'median', q3, 'sum']

    print("< Essays sentences feats >")    
    df['sent'] = df['essay'].apply(lambda x: re.split('\\.|\\?|\\!',x))
    df = df.explode('sent')
    df['sent'] = df['sent'].apply(lambda x: x.replace('\n','').strip())
    df['sent_len'] = df['sent'].apply(lambda x: len(x))
    df = df[df.sent_len!=0].reset_index(drop=True)

    sent_agg_df = df[['id','sent_len']].groupby(['id']).agg(AGGREGATIONS)
    sent_agg_df.columns = ['_'.join(x) for x in sent_agg_df.columns]
    sent_agg_df['id'] = sent_agg_df.index
    sent_agg_df = sent_agg_df.reset_index(drop=True)
    sent_agg_df = sent_agg_df.rename(columns={"sent_len_count":"sent_count"})
    return sent_agg_df


In [43]:
def essay_par_length(df):
    AGGREGATIONS = ['count', 'mean', 'min', 'max', 'first', 'last', q1, 'median', q3, 'sum']

    print("< Essays paragraphs feats >")    
    df['paragraph'] = df['essay'].apply(lambda x: x.split('\n'))
    df = df.explode('paragraph')
    df['paragraph_len'] = df['paragraph'].apply(lambda x: len(x)) 
    df = df[df.paragraph_len!=0].reset_index(drop=True)
    
    paragraph_agg_df = df[['id','paragraph_len']].groupby(['id']).agg(AGGREGATIONS)
                                 
    paragraph_agg_df.columns = ['_'.join(x) for x in paragraph_agg_df.columns]
    paragraph_agg_df['id'] = paragraph_agg_df.index
    paragraph_agg_df = paragraph_agg_df.reset_index(drop=True)
    paragraph_agg_df = paragraph_agg_df.rename(columns={"paragraph_len_count":"paragraph_count"})
    return paragraph_agg_df

def essay_par_words(df):
    AGGREGATIONS = ['count', 'mean', 'min', 'max', 'first', 'last', q1, 'median', q3, 'sum']
    print("< Essays paragraphs feats >")    
    df['paragraph'] = df['essay'].apply(lambda x: x.split('\n'))
    df = df.explode('paragraph')
    df['paragraph_word_count'] = df['paragraph'].apply(lambda x: len(x.split(' ')))
    
    paragraph_agg_df = df[['id','paragraph_word_count']].groupby(['id']).agg(AGGREGATIONS)
    paragraph_agg_df.columns = ['_'.join(x) for x in paragraph_agg_df.columns]
    paragraph_agg_df['id'] = paragraph_agg_df.index
    paragraph_agg_df = paragraph_agg_df.reset_index(drop=True)
    paragraph_agg_df.drop(columns=["paragraph_word_count_count"], inplace=True)
    paragraph_agg_df = paragraph_agg_df.rename(columns={"paragraph_len_count":"paragraph_count"})
    return paragraph_agg_df

< Essays paragraphs feats >


paragraph_count  paragraph_len_mean  paragraph_len_min  \
0                   3          508.000000                390   
1                   6          278.166667                176   
2                   6          429.500000                296   
3                   3          384.000000                347   
4                   5          283.400000                 23   
...               ...                 ...                ...   
2466                4          407.750000                301   
2467                6          387.166667                144   
2468                3          918.333333                327   
2469                5          509.600000                380   
2470                6          247.166667                 59   

      paragraph_len_max  paragraph_len_first  paragraph_len_last  \
0                   654                  390                 480   
1                   462                  240                 284   
2                   568                  491                 296   
3                   449                  347                 356   
4                   627                  351                  23   
...                 ...                  ...                 ...   
2466                514                  372                 301   
2467                648                  144                 228   
2468               2002                  426                2002   
2469                672                  672                 380   
2470                412                   59                 224   

      paragraph_len_q1  paragraph_len_median  paragraph_len_q3  \
0               435.00                 480.0            567.00   
1               228.75                 261.0            283.50   
2               356.75                 444.5            483.50   
3               351.50                 356.0            402.50   
4               124.00                 292.0            351.00   
...                ...                   ...               ...   
2466            354.25                 408.0            461.50   
2467            274.00                 424.0            450.25   
2468            376.50                 426.0           1214.00   
2469            394.00                 540.0            562.00   
2470            171.50                 229.5            358.00   

      paragraph_len_sum        id  
0                  1524  001519c8  
1                  1669  0022f953  
2                  2577  0042269b  
3                  1152  0059420b  
4                  1417  0075873a  
...                 ...       ...  
2466               1631  ffb8c745  
2467               2323  ffbef7e5  
2468               2755  ffccd6fd  
2469               2548  ffec5b38  
2470               1483  fff05981  

[2471 rows x 11 columns]

In [45]:
def essay_par_words(df):
    AGGREGATIONS = ['count', 'mean', 'min', 'max', 'first', 'last', q1, 'median', q3, 'sum']
    print("< Essays paragraphs feats >")    
    df['paragraph'] = df['essay'].apply(lambda x: x.split('\n'))
    df = df.explode('paragraph')
    df['paragraph_word_count'] = df['paragraph'].apply(lambda x: len(x.split(' ')))
    
    paragraph_agg_df = df[['id','paragraph_word_count']].groupby(['id']).agg(AGGREGATIONS)
    paragraph_agg_df.columns = ['_'.join(x) for x in paragraph_agg_df.columns]
    paragraph_agg_df['id'] = paragraph_agg_df.index
    paragraph_agg_df = paragraph_agg_df.reset_index(drop=True)
    paragraph_agg_df.drop(columns=["paragraph_word_count_count"], inplace=True)
    paragraph_agg_df = paragraph_agg_df.rename(columns={"paragraph_len_count":"paragraph_count"})
    return paragraph_agg_df
parag_feats(train_essays)

< Essays paragraphs feats >


paragraph_word_count_mean  paragraph_word_count_min  \
0                     54.200000                         1   
1                     50.857143                         1   
2                     37.727273                         1   
3                     69.333333                        62   
4                     28.888889                         1   
...                         ...                       ...   
2466                  77.000000                        63   
2467                  34.615385                         1   
2468                 264.285714                         1   
2469                  83.400000                        62   
2470                  24.900000                         1   

      paragraph_word_count_max  paragraph_word_count_first  \
0                          112                          71   
1                           96                          53   
2                           88                          79   
3                           81                          62   
4                          114                          61   
...                        ...                         ...   
2466                        88                          71   
2467                       119                          27   
2468                      1703                          83   
2469                       111                         111   
2470                        66                          10   

      paragraph_word_count_last  paragraph_word_count_q1  \
0                            86                      1.0   
1                             1                     41.5   
2                            45                      1.0   
3                            65                     63.5   
4                             3                      1.0   
...                         ...                      ...   
2466                         63                     69.0   
2467                          1                      1.0   
2468                       1703                      1.0   
2469                         62                     66.0   
2470                         38                      1.0   

      paragraph_word_count_median  paragraph_word_count_q3  \
0                            71.0                    86.00   
1                            53.0                    61.50   
2                            45.0                    73.50   
3                            65.0                    73.00   
4                             3.0                    52.00   
...                           ...                      ...   
2466                         78.5                    86.50   
2467                          1.0                    80.00   
2468                          1.0                    71.50   
2469                         85.0                    93.00   
2470                         15.5                    43.25   

      paragraph_word_count_sum        id  
0                          271  001519c8  
1                          356  0022f953  
2                          415  0042269b  
3                          208  0059420b  
4                          260  0075873a  
...                        ...       ...  
2466                       308  ffb8c745  
2467                       450  ffbef7e5  
2468                      1850  ffccd6fd  
2469                       417  ffec5b38  
2470                       249  fff05981  

[2471 rows x 10 columns]

In [41]:
essay_sent_length(train_essays)

< Essays sentences feats >


sent_count  sent_len_mean  sent_len_min  sent_len_max  sent_len_first  \
0             14     106.142857            31           196              31   
1             15     107.666667            19           226              19   
2             19     133.842105            73           189             139   
3             13      86.846154            39           144              99   
4             16      86.812500            22           182              75   
...          ...            ...           ...           ...             ...   
2466          13     121.076923            55           180              79   
2467          29      78.310345            20           175             143   
2468           4     277.000000           200           359             223   
2469          27      92.592593            36           176              79   
2470          11     133.272727            56           411              57   

      sent_len_last  sent_len_q1  sent_len_median  sent_len_q3  sent_len_sum  \
0                89        75.50            119.5       126.00          1486   
1               143        56.50             92.0       149.50          1615   
2               161       108.00            139.0       161.00          2543   
3                80        62.00             80.0        99.00          1129   
4                22        60.00             74.0       106.25          1389   
...             ...          ...              ...          ...           ...   
2466            109        84.00            132.0       147.00          1574   
2467             52        52.00             67.0       105.00          2271   
2468            359       217.25            274.5       334.25          1108   
2469             94        63.00             98.0       105.50          2500   
2470            223        69.00             95.0       145.00          1466   

            id  
0     001519c8  
1     0022f953  
2     0042269b  
3     0059420b  
4     0075873a  
...        ...  
2466  ffb8c745  
2467  ffbef7e5  
2468  ffccd6fd  
2469  ffec5b38  
2470  fff05981  

[2471 rows x 11 columns]

In [ ]:
train_essays = get_essay_df(train_logs.collect().to_pandas())
essay_sent_words(train_essays)

sent_word_count_mean  sent_word_count_max  sent_word_count_first  \
0                17.133333                   29                      6   
1                20.375000                   45                      3   
2                20.450000                   29                     21   
3                14.928571                   27                     17   
4                15.058824                   35                     11   
...                    ...                  ...                    ...   
2466             19.571429                   33                     16   
2467             14.800000                   33                     27   
2468             41.000000                   61                     42   
2469             14.964286                   29                     11   
2470             20.416667                   66                      9   

      sent_word_count_q1  sent_word_count_median  sent_word_count_q3  \
0                  11.00                    21.0               22.00   
1                  10.50                    18.5               30.50   
2                  17.00                    21.0               26.25   
3                  11.00                    14.5               17.75   
4                  11.00                    12.0               18.00   
...                  ...                     ...                 ...   
2466               15.25                    22.5               24.00   
2467                9.25                    13.0               19.00   
2468               41.00                    42.0               60.00   
2469               10.75                    14.5               18.25   
2470               10.50                    16.0               24.25   

      sent_word_count_sum        id  
0                     257  001519c8  
1                     326  0022f953  
2                     409  0042269b  
3                     209  0059420b  
4                     256  0075873a  
...                   ...       ...  
2466                  274  ffb8c745  
2467                  444  ffbef7e5  
2468                  205  ffccd6fd  
2469                  419  ffec5b38  
2470                  245  fff05981  

[2471 rows x 8 columns]

In [ ]:
best_feats = []

train_feats, test_feats = pd.DataFrame(), pd.DataFrame()
best_rmse = float('inf')
round = 0
added_feats = []
improved = True
results = pd.DataFrame()
train_feats = train_feats.sample(frac=1).reset_index(drop=True)
list_train_feats = [feat for feat in feat_list if feat.startswith('train_')]

while improved:
    print(f'Starting round {round} of training feats')
    improved = False
    list_train_feats = [feat for feat in list_train_feats if feat not in added_feats]

    for tr_feats_cand in list_train_feats:
        ts_feats_cand = tr_feats_cand.replace('train', 'test')

        tr_feats = pd.read_pickle(f'{FEAT_STORE}/{tr_feats_cand}')
        ts_feats = pd.read_pickle(f'{FEAT_STORE}/{ts_feats_cand}')

        existing_train_columns = set(train_feats.columns)
        existing_test_columns = set(test_feats.columns)

        if not(train_feats.empty & test_feats.empty):

            train_feats = train_feats.merge(tr_feats, on='id', how='left')
            test_feats = test_feats.merge(ts_feats, on='id', how='left')

            if 'score' not in train_feats.columns:
                train_feats = train_feats.merge(train_scores, on='id', how='left')
                
            assert train_feats.shape[1] == test_feats.shape[1] + 1
        else:
            #print(f'feats empty - setting up train_feats')
            train_feats = tr_feats
            test_feats = ts_feats
            train_feats = train_feats.merge(train_scores, on='id', how='left')

        print(f'Training... {tr_feats_cand}')
        # print(f'Train feats cols {train_feats.columns}')
        tr_cols = tr_feats.drop(columns=['id']).columns
        ts_cols = ts_feats.drop(columns=['id']).columns
        train_feats = train_feats.sample(frac=1).reset_index(drop=True)
        mean_rmse = []

        for i in range(15):
            train_feats = train_feats.sample(frac=1).reset_index(drop=True)
            test_preds, valid_preds, final_rmse, cv_rm = lgb_pipeline(train_feats, test_feats, param)
            mean_rmse.append(final_rmse)

        temp_res = {'feat_name': tr_feats_cand, 'RMSE': np.mean(mean_rmse)}
        results = pd.concat([results, pd.DataFrame([temp_res])])

        train_feats.drop(columns=tr_cols, inplace=True)
        test_feats.drop(columns=ts_cols, inplace=True)

    results = results.sort_values('RMSE', ascending=True)
    top_score = results.head(1).RMSE.values[0]
    top_feat = results.head(1).feat_name.values[0]

    if top_score < best_rmse:
        best_rmse = top_score
        best_feat = top_feat
        improved = True
        print(f'Results improved!: Selected feat {top_feat} - score {top_score}')

        ts_top_feat = top_feat.replace('train', 'test')
        tr_feats = pd.read_pickle(f'{FEAT_STORE}/{top_feat}')
        ts_feats = pd.read_pickle(f'{FEAT_STORE}/{ts_top_feat}')

        if round > 0:
            train_feats = train_feats.merge(tr_feats, on='id', how='left')
            test_feats = test_feats.merge(ts_feats, on='id', how='left')
        else:
            train_feats = tr_feats
            test_feats = ts_feats

        added_feats.append(top_feat)
        round += 1
    else:
        print('Training Over!')

    list_train_feats = [feat for feat in list_train_feats if feat not in added_feats]
    print(f'list_train_feats: {list_train_feats}')
    print(f'added_feats_list: {added_feats}')
    print(f'best feat: {top_feat}')

print(f"Best RMSE: {best_rmse:.4f}")
print(f"Best Feature Set: {added_feats}")
best_feats.append(added_feats)
added_feats = []

Starting round 0 of training feats
Training... train_categorical_nunique.pkl
Training... train_essay_sentences.pkl
Training... train_r_burst_feats.pkl
Training... train_action_time_by_activity.pkl
Training... train_word_count_time_based.pkl
Training... train_word_pauses.pkl
Training... train_cursor_pos_acceleration.pkl
Training... train_word_counts_rate_of_change.pkl
Training... train_create_pauses.pkl
Training... train_get_keys_pressed_per_second.pkl
Training... train_input_text_change_feats.pkl
Training... train_events_counts_acceleration.pkl
Training... train_sentence_pauses.pkl
Training... train_events_counts_rate_of_change.pkl
Training... train_p_burst_feats.pkl
Training... train_count_of_activities.pkl
Training... train_events_counts_baseline.pkl
Training... train_paragraph_pauses.pkl
Training... train_action_time_baseline_stats.pkl
Training... train_cursor_pos_time_based.pkl
Training... train_essay_words.pkl
Training... train_word_count_acceleration.pkl
Training... train_cursor_

In [21]:
sent_agg_df

sent_word_count_mean  sent_word_count_max  sent_word_count_first  \
0                17.133333                   29                      6   
1                20.375000                   45                      3   
2                20.450000                   29                     21   
3                14.928571                   27                     17   
4                15.058824                   35                     11   
...                    ...                  ...                    ...   
2466             19.571429                   33                     16   
2467             14.800000                   33                     27   
2468             41.000000                   61                     42   
2469             14.964286                   29                     11   
2470             20.416667                   66                      9   

      sent_word_count_q1  sent_word_count_median  sent_word_count_q3  \
0                  11.00                    21.0               22.00   
1                  10.50                    18.5               30.50   
2                  17.00                    21.0               26.25   
3                  11.00                    14.5               17.75   
4                  11.00                    12.0               18.00   
...                  ...                     ...                 ...   
2466               15.25                    22.5               24.00   
2467                9.25                    13.0               19.00   
2468               41.00                    42.0               60.00   
2469               10.75                    14.5               18.25   
2470               10.50                    16.0               24.25   

      sent_word_count_sum        id  
0                     257  001519c8  
1                     326  0022f953  
2                     409  0042269b  
3                     209  0059420b  
4                     256  0075873a  
...                   ...       ...  
2466                  274  ffb8c745  
2467                  444  ffbef7e5  
2468                  205  ffccd6fd  
2469                  419  ffec5b38  
2470                  245  fff05981  

[2471 rows x 8 columns]

In [18]:
sent_agg_df.sent_word_count_last.value_counts()

1      2385
2         7
5         7
10        5
3         5
15        5
9         4
7         4
4         4
27        3
22        3
12        3
21        3
39        2
8         2
6         2
17        2
26        2
40        1
91        1
153       1
138       1
75        1
18        1
28        1
37        1
32        1
23        1
14        1
13        1
46        1
44        1
29        1
52        1
107       1
34        1
24        1
25        1
33        1
42        1
31        1
Name: sent_word_count_last, dtype: int64

In [ ]:
def sent_feats(df):
    print("< Essays sentences feats >")    
    df['sent'] = df['essay'].apply(lambda x: re.split('\\.|\\?|\\!',x))
    df = df.explode('sent')
    df['sent'] = df['sent'].apply(lambda x: x.replace('\n','').strip())
    # Number of characters in sentences
    df['sent_len'] = df['sent'].apply(lambda x: len(x))
    # Number of words in sentences
    df['sent_word_count'] = df['sent'].apply(lambda x: len(x.split(' ')))
    df = df[df.sent_len!=0].reset_index(drop=True)

    sent_agg_df = pd.concat([df[['id','sent_len']].groupby(['id']).agg(AGGREGATIONS), 
                             df[['id','sent_word_count']].groupby(['id']).agg(AGGREGATIONS)], axis=1)
    sent_agg_df.columns = ['_'.join(x) for x in sent_agg_df.columns]
    sent_agg_df['id'] = sent_agg_df.index
    sent_agg_df = sent_agg_df.reset_index(drop=True)
    sent_agg_df.drop(columns=["sent_word_count_count"], inplace=True)
    sent_agg_df = sent_agg_df.rename(columns={"sent_len_count":"sent_count"})
    return sent_agg_df


def parag_feats(df):
    print("< Essays paragraphs feats >")    
    df['paragraph'] = df['essay'].apply(lambda x: x.split('\n'))
    df = df.explode('paragraph')
    # Number of characters in paragraphs
    df['paragraph_len'] = df['paragraph'].apply(lambda x: len(x)) 
    # Number of words in paragraphs
    df['paragraph_word_count'] = df['paragraph'].apply(lambda x: len(x.split(' ')))
    df = df[df.paragraph_len!=0].reset_index(drop=True)
    
    paragraph_agg_df = pd.concat([df[['id','paragraph_len']].groupby(['id']).agg(AGGREGATIONS), 
                                  df[['id','paragraph_word_count']].groupby(['id']).agg(AGGREGATIONS)], axis=1) 
    paragraph_agg_df.columns = ['_'.join(x) for x in paragraph_agg_df.columns]
    paragraph_agg_df['id'] = paragraph_agg_df.index
    paragraph_agg_df = paragraph_agg_df.reset_index(drop=True)
    paragraph_agg_df.drop(columns=["paragraph_word_count_count"], inplace=True)
    paragraph_agg_df = paragraph_agg_df.rename(columns={"paragraph_len_count":"paragraph_count"})
    return paragraph_agg_df

In [3]:
# everything is logged - DONE
# bursts = 2/3 of a second - input only - DONE
# inter word pauses
# between sentence pauses ?
# between paragraph pauses ?
# backspace pauses
# edit pauses

In [4]:
# # TEST INDIVIDUAL FEATURES
# from m3_model_params import lgb_params_1
# tr_word_pause, ts_word_pause = word_pauses(train_logs, test_logs)
# train_feats = tr_word_pause.join(train_scores, on='id', how='left')
# test_feats = ts_word_pause.clone()

# train_feats = train_feats.collect().to_pandas()
# test_feats = test_feats.collect().to_pandas()
# test_preds, valid_preds, final_rmse, cv_rm = lgb_pipeline(train_feats, test_feats, lgb_params_1)

In [5]:
# [('train_down_events_counts.pkl', 26),
#  ('train_vector_one_gram.pkl', 26),
#  ('train_create_pauses.pkl', 26),
#  ('train_essay_paragraphs.pkl', 26),
#  ('train_cursor_pos_acceleration.pkl', 11),
#  ('train_word_count_acceleration.pkl', 6),
#  ('train_p_burst_feats.pkl', 5),
#  ('train_r_burst_feats.pkl', 3),
#  ('train_events_counts_acceleration.pkl', 3),
#  ('train_essay_sentences.pkl', 3),
#  ('train_categorical_nunique.pkl', 3),
#  ('train_vector_two_gram.pkl', 2),
#  ('train_cursor_pos_rate_of_change.pkl', 2),
#  ('train_word_counts_rate_of_change.pkl', 2),
#  ('train_count_of_activities.pkl', 2),
#  ('train_action_time_by_activity.pkl', 1),
#  ('train_product_to_keys.pkl', 1),
#  ('train_IKI_based_fractals.pkl', 1),
#  ('train_events_counts_baseline.pkl', 1)]

In [6]:
# best_feature_set_1 - PARTIAL
train_essays          = get_essay_df(train_logs.collect().to_pandas())
test_essays           = get_essay_df(test_logs.collect().to_pandas())

tr_down_events_counts, ts_down_events_counts = down_events_counts(train_logs, test_logs)
tr_vect_one, ts_vect_one = countvectorize_one_one(train_essays, test_essays)
tr_pauses, ts_pauses = create_pauses(train_logs, test_logs)
tr_cursor_pos_acc, ts_cursor_pos_acc = cursor_pos_acceleration(train_logs, test_logs)
tr_word_pause, ts_word_pause = word_pauses(train_logs, test_logs)
tr_word_count_acc, ts_word_count_acc = word_count_acceleration(train_logs, test_logs)
#tr_p_burst, ts_p_burst = p_burst_feats(train_logs, test_logs, 2)
tr_r_burst, ts_r_burst = r_burst_feats(train_logs, test_logs)
#tr_event_acc, ts_event_acc = events_counts_acceleration(train_logs, test_logs)
# tr_nunique, ts_nunique = categorical_nunique(train_logs, test_logs)
tr_vect_two, ts_vect_two = countvectorize_two_one(train_essays, test_essays)
# tr_time_by_act, ts_time_by_act = action_time_by_activity(train_logs, test_logs)
# tr_cursor_pos_roc, ts_cursor_pos_roc = cursor_pos_rate_of_change(train_logs, test_logs)
# 
# tr_act_count, ts_act_count = count_of_activities(train_logs, test_logs)
# tr_get_keys, ts_get_keys = get_keys_pressed_per_second(train_logs.collect().to_pandas(), 
#                                                        test_logs.collect().to_pandas())
# 
# tr_input_change, ts_input_change = input_text_change_feats(train_logs, test_logs)
# tr_wc_roc, ts_wc_roc =  word_counts_rate_of_change(train_logs, test_logs)

train_feats = tr_down_events_counts.join(tr_vect_one, on='id', how='left')
train_feats = train_feats.join(tr_pauses, on='id', how='left')
train_feats = train_feats.join(tr_cursor_pos_acc, on='id', how='left')
train_feats = train_feats.join(tr_word_pause, on='id', how='left')
train_feats = train_feats.join(tr_word_count_acc, on='id', how='left')
#train_feats = train_feats.join(tr_p_burst, on='id', how='left')
train_feats = train_feats.join(tr_r_burst, on='id', how='left')
train_feats = train_feats.join(tr_vect_two, on='id', how='left')
# train_feats = train_feats.join(tr_event_acc, on='id', how='left')
# train_feats = train_feats.join(tr_nunique, on='id', how='left')
# train_feats = train_feats.join(tr_wc_roc, on='id', how='left')
# train_feats = train_feats.join(tr_act_count, on='id', how='left')
# train_feats = train_feats.join(tr_cursor_pos_roc, on='id', how='left')

# train_feats = train_feats.join(tr_get_keys, on='id', how='left')
# train_feats = train_feats.join(tr_input_change, on='id', how='left')
# train_feats = train_feats.join(tr_time_by_act, on='id', how='left')

test_feats = ts_down_events_counts.join(ts_vect_one, on='id', how='left')
test_feats = test_feats.join(ts_pauses, on='id', how='left')
test_feats = test_feats.join(ts_cursor_pos_acc, on='id', how='left')
test_feats = test_feats.join(ts_word_pause, on='id', how='left')
test_feats = test_feats.join(ts_word_count_acc, on='id', how='left')
# test_feats = test_feats.join(ts_p_burst, on='id', how='left')
test_feats = test_feats.join(ts_r_burst, on='id', how='left')
test_feats = test_feats.join(ts_vect_two, on='id', how='left')
# test_feats = test_feats.join(tr_event_acc, on='id', how='left')
# test_feats = test_feats.join(ts_nunique, on='id', how='left')
# test_feats = test_feats.join(ts_wc_roc, on='id', how='left')
# test_feats = test_feats.join(ts_act_count, on='id', how='left')
# test_feats = test_feats.join(ts_cursor_pos_roc, on='id', how='left')


# test_feats = test_feats.join(ts_get_keys, on='id', how='left')
# test_feats = test_feats.join(ts_input_change, on='id', how='left')
# test_feats = test_feats.join(ts_time_by_act, on='id', how='left')


train_logs = train_logs.collect().to_pandas()
test_logs = test_logs.collect().to_pandas()
train_scores = train_scores.collect().to_pandas()
train_feats = train_feats.sort('id')
train_feats = train_feats.collect().to_pandas()
test_feats = test_feats.collect().to_pandas()

train_feats           = train_feats.merge(parag_feats(train_essays), on='id', how='left')
test_feats            = test_feats.merge(parag_feats(test_essays), on='id', how='left')

train_feats           = train_feats.merge(sent_feats(train_essays), on='id', how='left')
test_feats            = test_feats.merge(sent_feats(test_essays), on='id', how='left')

train_feats = train_feats.merge(train_scores, on='id', how='left')
print(f'train feats shape {train_feats.shape}')


< Events counts features >
< Count vectorize one-grams >
< Idle time features >
< cursor position acceleration >
word pauses
< word count acceleration >
< R-burst features >
< Count vectorize bi-grams >
< Essays paragraphs feats >
< Essays paragraphs feats >
< Essays sentences feats >
< Essays sentences feats >
train feats shape (2471, 144)


In [7]:
from m5_sb_models import lgb_pipeline
lgb_params_1 = {
    'boosting_type': 'gbdt', 
    'metric': 'rmse',
    'reg_alpha': 0.0031, 
    'reg_lambda': 0.001, 
    'colsample_bytree': 0.8,  
    'subsample_freq': 1,  
    'subsample': 0.75,  
    'learning_rate': 0.017, 
    'num_leaves': 19, 
    'min_child_samples': 46,
    'n_estimators': 350,
    'verbosity': -1
    }

param = {'n_estimators': 1024,
        'learning_rate': 0.005,
        'metric': 'rmse',
        'force_col_wise': True,
        'verbosity': 0,}

# train_feats = train_feats[['id', 'score'] + feat_select]
# test_feats = test_feats[['id'] + feat_select]

print(f'train feats shape {train_feats.shape}')


train feats shape (2471, 144)


In [8]:
shuffle_preds = []

for i in range(15):
    train_feats = train_feats.sample(frac=1).reset_index(drop=True)
    test_preds, oof_preds, rmse, model = lgb_pipeline(train_feats, test_feats, lgb_params_1)
    shuffle_preds.append(rmse)
    #test_preds, oof_preds, rmse, model = lgb_pipeline(train_feats, test_feats, param)

np.mean(shuffle_preds)

Final RMSE over 50: 0.603617. Std 0.8270
RMSE by fold 0.603563. Std 0.0083
Final RMSE over 50: 0.601545. Std 0.8255
RMSE by fold 0.601473. Std 0.0091
Final RMSE over 50: 0.602574. Std 0.8276
RMSE by fold 0.602372. Std 0.0151
Final RMSE over 50: 0.601950. Std 0.8275
RMSE by fold 0.601897. Std 0.0080
Final RMSE over 50: 0.604154. Std 0.8274
RMSE by fold 0.603977. Std 0.0149
Final RMSE over 50: 0.602163. Std 0.8266
RMSE by fold 0.602061. Std 0.0108
Final RMSE over 50: 0.603044. Std 0.8271
RMSE by fold 0.602868. Std 0.0148
Final RMSE over 50: 0.602455. Std 0.8275
RMSE by fold 0.602338. Std 0.0117
Final RMSE over 50: 0.602454. Std 0.8279
RMSE by fold 0.602405. Std 0.0079
Final RMSE over 50: 0.602837. Std 0.8266
RMSE by fold 0.602635. Std 0.0155
Final RMSE over 50: 0.602299. Std 0.8262
RMSE by fold 0.602235. Std 0.0083
Final RMSE over 50: 0.602076. Std 0.8286
RMSE by fold 0.601953. Std 0.0120
Final RMSE over 50: 0.602816. Std 0.8265
RMSE by fold 0.602714. Std 0.0110
Final RMSE over 50: 0.602

0.6026620396931698

In [9]:
#  0.6046250116470197 - baseline
#  0.6034392851713597 - word_pauses
#  0.6031532703206033 - word_count_acc
#  0.6030545803940613 - r_burst
#  0.60246384515886   - bigrams
# sentences

- M4 + cursor_pos_acc + word_count_acc + bigrams <br />
0.6044949502480581


- M4 + cursor_pos_acc + word_count_acc  <br />
0.6047963953877546 <br />


In [10]:
oof_res = oof_preds.groupby(['id', 'score'])['preds'].mean().reset_index()
# oof_res['rmse'] = oof_res.apply(lambda x: np.sqrt((x['score']-x['preds'])**2))
oof_res['RMSE'] = np.sqrt((oof_res['score']-oof_res['preds'])**2)
oof_res.groupby(['score'])['RMSE'].mean().reset_index().sort_values('RMSE', ascending=False)

score      RMSE
0     0.5  1.528224
1     1.0  1.283878
11    6.0  1.132529
2     1.5  0.879977
10    5.5  0.738216
3     2.0  0.573584
9     5.0  0.474352
4     2.5  0.461494
5     3.0  0.460465
6     3.5  0.402527
8     4.5  0.355785
7     4.0  0.349586

In [11]:
oof_res = oof_preds.groupby(['id', 'score'])['preds'].mean().reset_index()
oof_res['RMSE'] = np.sqrt((oof_res['score']-oof_res['preds'])**2)
oof_res.groupby(['score'])['RMSE'].mean().reset_index().sort_values('RMSE', ascending=False)

score      RMSE
0     0.5  1.528224
1     1.0  1.283878
11    6.0  1.132529
2     1.5  0.879977
10    5.5  0.738216
3     2.0  0.573584
9     5.0  0.474352
4     2.5  0.461494
5     3.0  0.460465
6     3.5  0.402527
8     4.5  0.355785
7     4.0  0.349586

In [12]:
oof_res = oof_preds.groupby(['id', 'score'])['preds'].mean().reset_index()
# oof_res['rmse'] = oof_res.apply(lambda x: np.sqrt((x['score']-x['preds'])**2))
oof_res['RMSE'] = np.sqrt((oof_res['score']-oof_res['preds'])**2)

In [13]:
import lightgbm as lgb

weights = [1.0 for _ in range(len(y))]  # Default weights
weights[some_specific_index] = 1.5  # Higher weight for a specific instance

train_data = lgb.Dataset(X_train, label=y_train, weight=weights)
# Proceed with setting up parameters and training the model


NameError: name 'y' is not defined

train feats shape (2471, 151)
Final RMSE over 50: 0.605768. Std 0.8244
RMSE by fold 0.605635. Std 0.0128

reduced feats
Final RMSE over 50: 0.604472. Std 0.8245
RMSE by fold 0.604365. Std 0.0115

r burst only
Number of estimators: 350
Final RMSE over 50: 0.604057. Std 0.8244
RMSE by fold 0.603946. Std 0.0116

Number of estimators: 350
Final RMSE over 50: 0.604650. Std 0.8243
RMSE by fold 0.604527. Std 0.0122

In [ ]:
test_ids = test_feats.id
y_pred = np.mean(test_preds, axis=0)

sub = pd.DataFrame({'id': test_ids, 'score': y_pred})
sub.to_csv('submission.csv', index=False)